LG-9855 delete ssn

svalexander opened the pull request:
## 🎫 Ticket

[LG-9855](https://cm-jira.usa.gov/browse/LG-9855)


## 🛠 Summary of changes

Remove old flow state machine version of the ssn page. 


## 📜 Testing Plan

Provide a checklist of steps to confirm the changes.

- [ ] Walk through ipp flow and arrive at ssn page
- [ ] Make sure url is `in_person_proofing/ssn`
- [ ] Try updating url to `in_person/ssn`, it should redirect to `in_person_proofing/ssn`
- [ ] Make sure cancel link appears at bottom of page and works (opt not to cancel to go back to ssn page)
- [ ] Continue to verify_info page and make sure ssn is correct
- [ ] Click update for ssn and verify you're taken back to ssn page
- [ ] Make sure back button appears at bottom of page and works
- [ ] Walk through remote flow w/ a successful id --> we just want to make sure this flow isn't changed
- [ ] Make sure url is `verify/ssn`
- [ ] Make sure cancel link appears at bottom of page and works (opt not to cancel to go back to ssn page)
- [ ] Continue to verify_info page and make sure ssn is correct
- [ ] Click update for ssn and verify you're taken back to ssn page
- [ ] Make sure back button appears at bottom of page and works

soniaconnolly commented on app/controllers/idv/in_person/verify_info_controller.rb:
```diff
@@ -51,11 +51,7 @@ def invalid_state?
-        else
-          idv_in_person_step_url(step: :ssn)
-        end
+        idv_in_person_ssn_url
```
I think it might be better to merge this PR and then change the url in a separate PR. It's worth thinking through whether this will work though.

svalexander replied:
we're going to follow this suggestion; added a [ticket](https://cm-jira.usa.gov/browse/LG-10586)

gina-yamada replied:
Make sure to update your testing plan. If a different issue will handle changing the url, the following is no longer accurate. 
```
 Make sure url is in_person/ssn
 Try updating url to in_person_proofing/ssn, it should display a 404
 ```

soniaconnolly replied:
When you do change the url, `in_person_proofing/ssn` should redirect to `in_person/ssn` rather than 404 so that the change is invisible to the user.

soniaconnolly commented on app/forms/idv/ssn_form.rb:
```diff
@@ -53,4 +53,4 @@ def raise_invalid_ssn_parameter_error(key)
     end
   end
-end
+  end
```
Spacing typo
```suggestion
end
```

soniaconnolly commented on config/routes.rb:
```diff
@@ -382,8 +382,8 @@
-      get '/in_person_proofing/ssn' => 'in_person/ssn#show'
-      put '/in_person_proofing/ssn' => 'in_person/ssn#update'
+      get '/in_person/ssn' => 'in_person/ssn#show'
+      put '/in_person/ssn' => 'in_person/ssn#update'
```
In the PR where you change this url, suggest adding redirects under deprecated routes (near line 409):
      get '/in_person_proofing/ssn', to: redirect('/verify/in_person/ssn')
      put '/in_person_proofing/ssn', to: redirect('/verify/in_person/ssn')

soniaconnolly commented on spec/controllers/idv/in_person/ssn_controller_spec.rb:
```diff
@@ -120,21 +71,8 @@
       expect(@analytics).to have_received(:track_event).with(analytics_name, analytics_args)
     end
 
-    it 'updates DocAuthLog ssn_view_count' do
-      doc_auth_log = DocAuthLog.create(user_id: user.id)
-
-      expect { get :show }.to(
-        change { doc_auth_log.reload.ssn_view_count }.from(0).to(1),
-      )
-    end
```
We're still using the DocAuthLog counts. They are separate from the step counts in the flow_session that were deleted.

soniaconnolly commented on spec/controllers/idv/in_person/ssn_controller_spec.rb:
```diff
@@ -120,21 +71,8 @@
-
     context 'with an ssn in session' do
-      let(:referer) { idv_in_person_step_url(step: :address) }
+      let(:referer) { idv_document_capture_url }
```
This change looks unexpected since the in person ssn step comes after the in person address step.

svalexander replied:
good point

soniaconnolly commented on spec/controllers/idv/in_person/ssn_controller_spec.rb:
```diff
@@ -120,21 +71,8 @@
-      )
-    end
-
-    it 'adds a session id to flow session' do
-      get :show
-      expect(flow_session[:threatmetrix_session_id]).to_not eq(nil)
-    end
```
Since this controller still has threatmetrix, this spec should probably still be included.

svalexander replied:
good catch!

soniaconnolly commented on spec/controllers/idv/in_person/ssn_controller_spec.rb:
```diff
@@ -24,75 +24,26 @@
     allow(subject).to receive(:pii_from_user).and_return(pii_from_user)
     allow(subject).to receive(:flow_session).and_return(flow_session)
     stub_sign_in(user)
-    subject.user_session['idv/in_person'] = flow_session
```
Is this still needed?

svalexander replied:
running the test with and without it both work so my thought is it might not really be needed

soniaconnolly commented on spec/controllers/idv/in_person/ssn_controller_spec.rb:
```diff
@@ -160,106 +98,96 @@
-          expect(flow_session[:pii_from_user][:ssn]).to eq(ssn)
-        end
+      it 'merges ssn into pii session value' do
+        put :update, params: params
 
-        it 'invalidates steps after ssn' do
-          put :update, params: params
+        expect(flow_session[:pii_from_user][:ssn]).to eq(ssn)
+      end
 
-          expect(subject.idv_session.applicant).to be_blank
-          expect(subject.idv_session.resolution_successful).to be_blank
-          expect(subject.idv_session.profile_confirmation).to be_blank
-          expect(subject.idv_session.vendor_phone_confirmation).to be_blank
-          expect(subject.idv_session.user_phone_confirmation).to be_blank
-        end
+      it 'invalidates steps after ssn' do
+        put :update, params: params
 
-        it 'redirects to the expected page' do
-          put :update, params: params
+        expect(subject.idv_session.applicant).to be_blank
+        expect(subject.idv_session.resolution_successful).to be_blank
+        expect(subject.idv_session.profile_confirmation).to be_blank
+        expect(subject.idv_session.vendor_phone_confirmation).to be_blank
+        expect(subject.idv_session.user_phone_confirmation).to be_blank
+      end
 
-          expect(response).to redirect_to idv_in_person_verify_info_url
-        end
+      it 'redirects to the expected page' do
+        put :update, params: params
 
-        it 'does not change threatmetrix_session_id when updating ssn' do
-          flow_session[:pii_from_user][:ssn] = ssn
-          put :update, params: params
-          session_id = flow_session[:threatmetrix_session_id]
-          subject.threatmetrix_view_variables
-          expect(flow_session[:threatmetrix_session_id]).to eq(session_id)
-        end
```
Probably still needed.

svalexander replied:
adding back in the threatmetrix test

soniaconnolly reviewed (commented):
Need to change final_url in in_person_controller to idv_in_person_ssn_url (or idv_in_person_proofing_ssn_url if you leave the url unchanged for now).

Check for remaining instances of Idv::Steps::InPerson::SsnStep and delete.

racingspider reviewed (approved):
lgtm

gina-yamada commented:
@svalexander  We will now need to also delete the code I added in this [PR](https://github.com/18F/identity-idp/pull/8968) - line 17 `flow_session['Idv::Steps::InPerson::SsnStep'] = true` as we will no longer have the FSM SSN step

gina-yamada commented:
**Notes/Observations for Review:**
- No instances of feature flag `IdentityConfig.store.in_person_ssn_info_controller_enabled` or `in_person_ssn_info_controller_enabled` found
- Old view deleted `app/views/idv/in_person/ssn.html.erb`
- Remove :step_count from this step’s analytics in spec/features/idv/analytics_spec.rb (updated also to include ab_test_analytics_buckets- which I'd expect)
- Remove step from Flow (no instances of idv_in_person_step_url(step: :ssn) observed)
- update final_url in Controller - observed here `app/controllers/idv/in_person_controller.rb`

still checking...
Delete old step and specs
Delete spec helper method for old step path and usage
Delete anywhere the step is marked complete or incomplete
Put up a PR to run the full test suite in GitLab, and add commits to fix failing specs
Some delete PRs are long and complex ([DocumentCaptureStep](https://github.com/18F/identity-idp/pull/8218/commits)) and some are simpler ([LinkSentStep](https://github.com/18F/identity-idp/pull/8394/commits))

gina-yamada commented on spec/features/idv/steps/in_person/verify_info_spec.rb:
```diff
@@ -25,7 +25,6 @@
     expect_in_person_step_indicator_current_step(t('step_indicator.flows.idv.verify_info'))
     expect(page).to have_current_path(idv_in_person_verify_info_path)
     expect(page).to have_content(t('headings.verify'))
-    expect(page).to have_current_path(idv_in_person_verify_info_path)
```
Good eye! I wondered why this was being deleted. I ran it to ensure it was not causing a failure when it was present.  It is a dup from line 26. 👏 👏  Good clean up

gina-yamada reviewed (changes requested):
delete the one instance of flow_session['Idv::Steps::InPerson::SsnStep'] = true in code

gina-yamada commented:
@svalexander  What do you think of adding a do not merge label on this PR?

svalexander commented:
I added the additional ac of changing final_url to this [ticket](https://cm-jira.usa.gov/browse/LG-10586), and removed the ref. to the ssnstep in this pr

gina-yamada commented on app/controllers/idv/in_person_controller.rb:
```diff
@@ -16,7 +16,7 @@ class InPersonController < ApplicationController
     FLOW_STATE_MACHINE_SETTINGS = {
       step_url: :idv_in_person_step_url,
-      final_url: :idv_in_person_proofing_ssn_url,
+      final_url: :idv_in_person_verify_info_url,
```
@svalexander  I think this should be `idv_in_person_proofing_ssn_url`. Check out my delete [PR](https://github.com/18F/identity-idp/pull/8534/files) for Verify. The final url is actually the first non FSM step

svalexander replied:
yes, you're right! i updated

gina-yamada commented:
I pulled in main. While checks ran, I did some manual testing locally. New SSN controller behaving as I'd expect! I will also notify deployer about a shorter 50/50 state is better for the next release.
